core/src/main/java/tc/oc/pgm/map/source/gitSource/GitMapSourceFactory.java

```diff
@@ -0,0 +1,149 @@
+package tc.oc.pgm.map.source.gitSource;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+import org.eclipse.jgit.api.CloneCommand;
+import org.eclipse.jgit.api.FetchCommand;
+import org.eclipse.jgit.api.Git;
+import org.eclipse.jgit.api.ResetCommand;
+import org.eclipse.jgit.api.errors.GitAPIException;
+import org.eclipse.jgit.api.errors.TransportException;
+import org.eclipse.jgit.transport.CredentialsProvider;
+import org.eclipse.jgit.transport.URIish;
+import org.eclipse.jgit.transport.UsernamePasswordCredentialsProvider;
+import tc.oc.pgm.api.map.MapSource;
+import tc.oc.pgm.api.map.exception.MapMissingException;
+import tc.oc.pgm.map.source.SystemMapSourceFactory;
+
+public class GitMapSourceFactory extends SystemMapSourceFactory {
+
+  private final URIish gitURI;
+  private final Logger logger;
+  private final File dir;
+  private final GitRepository repository;
+
+  private CredentialsProvider provider = null;
+
+  public GitMapSourceFactory(GitRepository repo, Logger logger) throws MapMissingException {
+    super(repo.getDir().getPath());
+
+    this.repository = repo;
+    this.gitURI = repository.getUri();
+
+    this.logger = logger;
+    this.dir = repo.getDir();
+
+    File masterDir = new File("./maps");
+    if (!masterDir.exists()) masterDir.mkdir();
+
+    // Ensure the git map directory exists before trying to update it
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+    if (!dir.isDirectory())
+      throw new MapMissingException(
+          dir.getPath(),
+          "Unable to read git directory " + gitURI.getHumanishName() + "(Is it a file?)");
+
+    if (repo.getUri().getPass() != null) {
+      provider =
+          new UsernamePasswordCredentialsProvider(repo.getUri().getUser(), repo.getUri().getPass());
+    }
+  }
+
+  @Override
+  public Iterator<? extends MapSource> loadNewSources() {
+
+    return new Iterator<MapSource>() {
+      private Iterator<MapSource> delegate;
+
+      @Override
+      public boolean hasNext() {
+        if (delegate == null) {
+          try {
+            refreshRepo();
+          } catch (GitAPIException e) {
+            logger.log(
+                Level.WARNING,
+                "Unable to fetch repo " + gitURI.getPath() + ", caused by: " + e.getMessage());
+            e.printStackTrace();
+            if (e instanceof TransportException) {
+              logger.log(
+                  Level.INFO, "Wrong username and password provided for " + gitURI.getPath());
+            }
+          } catch (MapMissingException e) {
+            e.printStackTrace();
+          }
```

Comment:
Although I said originally to make it lazy, this is probably too weird. So just git pull when initializing the `GitMapSourceFactory` and re-pull it before calling `loadNewSources`

Reply:
What do you mean "before"? On `/loadnewmaps` it only calls the `loadNewSources()`.

Reply:
Do you mean do something like this?:
```
@Override
public Iterator loadNewSources(){
refreshRepo();
super.loadNewSources();
}

```

Reply:
Yes

Reply:
Then i see no reason to also pull when initializing the `GitMapSourceFactory`, why twice?

Reply:
You only need to initialize or pull when `loadnewsources` is first loaded